src/workerd/api/streams/writable.c++

```diff
@@ -365,6 +365,154 @@ private:
   }
 };
 
+// In order to support JavaScript-backed WritableStreams that do not have a backing
+// WritableStreamSink, we need an alternative version of the WritableStreamRpcAdapter
+// that will arrange to acquire the isolate lock when necessary to perform writes
+// directly on the WritableStreamController. Note that this approach is necessarily
+// a lot slower
+class WritableStreamJsRpcAdapter final: public capnp::ExplicitEndOutputStream {
+public:
+  WritableStreamJsRpcAdapter(IoContext& context, jsg::Ref<WritableStreamDefaultWriter> writer)
+      : context(context), writer(kj::mv(writer)) {}
+
+  ~WritableStreamJsRpcAdapter() noexcept(false) {
+    weakRef->invalidate();
+    doneFulfiller->fulfill();
+  }
+
+  // Returns a promise that resolves when the stream is dropped. If the promise is canceled before
+  // that, the stream is revoked.
+  kj::Promise<void> waitForCompletionOrRevoke() {
+    auto paf = kj::newPromiseAndFulfiller<void>();
+    doneFulfiller = kj::mv(paf.fulfiller);
+
+    return paf.promise.attach(kj::defer([weakRef = weakRef->addRef()]() mutable {
+      KJ_IF_SOME(obj, weakRef->tryGet()) {
+        // Stream is still alive, revoke it.
+        auto exception = cancellationException();
+        if (!obj.canceler.isEmpty()) {
+          obj.canceler.cancel(kj::cp(exception));
+        }
+        KJ_DEFER(obj.writer = kj::none);
+        KJ_IF_SOME(writer, obj.writer) {
+          obj.context.addTask(obj.context.run(
+              [writer=kj::mv(writer), exception=kj::mv(exception)](Worker::Lock& lock) mutable {
+            jsg::Lock& js = lock;
+            auto ex = js.exceptionToJs(kj::mv(exception));
+            return IoContext::current().awaitJs(lock,
+                writer->abort(lock, ex.getHandle(js)));
+          }));
```

Comment:
For review: This part is a bit weird currently. We need to be able to trigger the abort algorithm on the WritableStream if we're still actually able to. At this point, however, the context might not exist anymore I think? Does the adapter need to have an explicit mechanism for triggering the `abort()` within the lock?

Reply:
I think that it is _always_ the case, in this branch, that the IoContext doesn't exist anymore:

* In the case where `waitForCompletionOrRevoke()` completes normally, then `weakRef` is already invalidated before this `kj::defer` callback runs, so none of this code executes.
* So, this block only executes in the case that `waitForCompletionOrRevoke()` was actually canceled prematurely.
* The promise is passed directly to `ioctx.addTask()`, so it would only be canceled if the `IoContext` is destroyed.
* Therefore, when executing this code, the `IoContext` is always destroyed (or in the process of being destroyed).

So I think it's unsafe to call `obj.context.run()` here. But also, you don't need to! Since the context is shutting down anyway, all I/O is canceled, including the abort algorithm.

Reply:
I think the place where you actually want to invoke the abort algorithm is in the destructor `~WritableStreamJsRpcAdapter()`, if `writer` is non-null at that point and `end()` was never called.